### PR TITLE
[AArch64] Improve frame code, implement emitForeignCall

### PR DESCRIPTION
**Pull request by christhalinger**

### CODE REVIEW

File: graal/com.oracle.graal.asm.aarch64/src/com/oracle/graal/asm/aarch64/AArch64MacroAssembler.java [hunk 1]
```diff
@@ -1325,13 +1332,23 @@ protected void patchJumpTarget(int branch, int jumpTarget) {
...
                 }
                 break;
+            }
+            case ADR: {
+                int information = instruction >>> PatchLabelKind.INFORMATION_OFFSET;
+                int regEncoding = information;
+                Register reg = AArch64.cpuRegisters[regEncoding];
+                // XXX The -4 is an ugly hack for the fact that we can't pass the Label to
+                // AArch64Call.directCall
+                super.adr(reg, branchOffset - 4, branch);
```
Comment (thomaswue) [inline graal/com.oracle.graal.asm.aarch64/src/com/oracle/graal/asm/aarch64/AArch64MacroAssembler.java]: Can we still avoid "4" as a constant here and have a constant instead?

File: graal/com.oracle.graal.hotspot.aarch64/src/com/oracle/graal/hotspot/aarch64/AArch64HotSpotBackend.java [hunk 2]
```diff
@@ -140,19 +140,11 @@ public void enter(CompilationResultBuilder crb) {
...
-                masm.str(64, lr, address);
-                masm.str(64, fp, address);
-                // Update framepointer
-                masm.mov(64, fp, scratch);
-
-                if (ZapStackOnMethodEntry.getValue()) {
+            if (ZapStackOnMethodEntry.getValue()) {
+                try (ScratchRegister sc = masm.getScratchRegister()) {
+                    Register scratch = sc.getRegister();
                     int intSize = 4;
```
Comment (thomaswue) [inline graal/com.oracle.graal.hotspot.aarch64/src/com/oracle/graal/hotspot/aarch64/AArch64HotSpotBackend.java]: That is Integer.SIZE >> 3?

File: graal/com.oracle.graal.lir.aarch64/src/com/oracle/graal/lir/aarch64/AArch64ArithmeticOp.java [hunk 3]
```diff
@@ -178,7 +177,12 @@ public void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm) {
                     masm.sub(size, dst, src, (int) b.asLong());
                     break;
                 case AND:
-                    masm.and(size, dst, src, b.asLong());
+                    // XXX Should this be handled somewhere else?
+                    if (size == 32 && b.asLong() == 0xFFFF_FFFFL) {
```
Comment (thomaswue) [inline graal/com.oracle.graal.lir.aarch64/src/com/oracle/graal/lir/aarch64/AArch64ArithmeticOp.java]: 32 == Integer.SIZE?

File: graal/com.oracle.graal.asm.aarch64/src/com/oracle/graal/asm/aarch64/AArch64MacroAssembler.java (hunk 1 above)
Comment (christhalinger) [inline graal/com.oracle.graal.asm.aarch64/src/com/oracle/graal/asm/aarch64/AArch64MacroAssembler.java]: This needs to go.  I discussed with @tkrodriguez to find a better solution.

File: graal/com.oracle.graal.lir.aarch64/src/com/oracle/graal/lir/aarch64/AArch64ArithmeticOp.java (hunk 3 above)
Comment (christhalinger) [inline graal/com.oracle.graal.lir.aarch64/src/com/oracle/graal/lir/aarch64/AArch64ArithmeticOp.java]: This is a hack and also needs to go.  Just need to find the right place to fix this.

File: graal/com.oracle.graal.hotspot.aarch64/src/com/oracle/graal/hotspot/aarch64/AArch64HotSpotBackend.java (hunk 2 above)
Comment (christhalinger) [inline graal/com.oracle.graal.hotspot.aarch64/src/com/oracle/graal/hotspot/aarch64/AArch64HotSpotBackend.java]: This is just copy-paste from AMD64:

                if (ZapStackOnMethodEntry.getValue()) {
                    final int intSize = 4;

Comment (christhalinger) [inline graal/com.oracle.graal.hotspot.aarch64/src/com/oracle/graal/hotspot/aarch64/AArch64HotSpotBackend.java]: ...and it should be `Integer.BYTES`.

**Comment by dougxc** [comment]
Sorry for prematurely closing this PR. @christhalinger I assume you'll open a follow-up PR to address the comments.

**Comment by christhalinger** [comment]
I'll address the comments by fixing them properly, yes.

**Pull request merged**